File: FRONTEND/src/services/productService.ts
import api from './api';
import { Product, ProductRequest } from '../types/product';

const PRODUCT_API_URL = '/api/products'; // Routed via API Gateway

export const productService = {
  getAllProducts: async () => {
    const response = await api.get<Product[]>(PRODUCT_API_URL);
    return response.data;
  },

  getProductById: async (id: number) => {
    const response = await api.get<Product>(`${PRODUCT_API_URL}/${id}`);
    return response.data;
  },

  createProduct: async (data: ProductRequest) => {
    const response = await api.post<Product>(PRODUCT_API_URL, data);
    return response.data;
  },

  updateProduct: async (id: number, data: ProductRequest) => {
    const response = await api.put<Product>(`${PRODUCT_API_URL}/${id}`, data);
    return response.data;
  },

  deleteProduct: async (id: number) => {
    await api.delete(`${PRODUCT_API_URL}/${id}`);
  },
};

File: BACKEND/product-service/src/main/java/com/cardpro/productservice/service/ProductService.java
package com.cardpro.productservice.service;

import com.cardpro.productservice.dto.ProductRequest;
import com.cardpro.productservice.dto.ProductResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface ProductService {

    ProductResponse createProduct(ProductRequest request);

    ProductResponse getProductById(Long id);

    List<ProductResponse> getAllProducts();

    Page<ProductResponse> getAllProducts(Pageable pageable);

    Page<ProductResponse> searchProducts(String keyword, Pageable pageable);

    ProductResponse updateProduct(Long id, ProductRequest request);

    void deleteProduct(Long id);
}